lib/fs/exfat.inc

```diff
@@ -0,0 +1,665 @@
+		; inputs
+		; start-cluster = first cluster in chain
+		; outputs
+		; t if successful
+		(when (>= start-cluster 2)
+			(defq current-cluster start-cluster)
+			(while (and current-cluster (>= current-cluster 2) (< current-cluster +exfat_fat_eoc))
+				(defq next-cluster (. this :read-fat-entry current-cluster))
+				(. this :write-fat-entry current-cluster +exfat_fat_free)
+				(setq current-cluster next-cluster)))
+		t)
+
+	;;;;;;;;;;;;;;;;;;;;
+	; Filesystem Operations
+	;;;;;;;;;;;;;;;;;;;;
+
+	(defmethod :format (size)
+		; Format the filesystem with the given size
+		; inputs
+		; size = total size in bytes
+		; outputs
+		; t if successful, nil otherwise
+		(defq stream (get this :stream)
+			sector-size (get this :sector-size)
+			cluster-shift (get this :cluster-shift)
+			cluster-size (get this :cluster-size)
+			total-sectors (/ size sector-size)
+			fat-offset +exfat_fat_offset)
+
+		; Calculate FAT size (32-bit entries, 1 entry per cluster)
+		(defq cluster-heap-start (+ fat-offset 128)  ; Reserve 64KB for FAT
+			cluster-count (/ (- total-sectors cluster-heap-start) (>> cluster-size (logn 2 sector-size)))
+			fat-length (/ (+ (* cluster-count 4) sector-size -1) sector-size))  ; Round up
+
+		; Update filesystem parameters
+		(setd this :fat-offset fat-offset
+			:fat-length fat-length
+			:cluster-heap-offset (+ fat-offset fat-length)
+			:cluster-count cluster-count
+			:root-dir-cluster 2
+			:mounted nil)
+
+		; Write boot sector
+		(. this :write-boot-sector size)
+
+		; Initialize FAT (all entries free except reserved and root)
+		(each! 0 fat-length
+			(lambda (i)
+				(. this :write-sector (+ fat-offset i) (str-alloc sector-size))))
+
+		; Mark first two FAT entries as reserved
+		(. this :write-fat-entry 0 0xFFFFFFF8)  ; Media type
+		(. this :write-fat-entry 1 +exfat_fat_eoc)
+
+		; Mark root directory cluster as end of chain
+		(. this :write-fat-entry 2 +exfat_fat_eoc)
+
+		; Initialize root directory (empty)
+		(. this :write-cluster 2 (cat
+			; Allocation bitmap entry
+			(. this :make-alloc-bitmap-entry)
+			; End of directory marker
+			(str-alloc (- cluster-size +exfat_entry_size))))
+
+		(setd this :mounted t)
+		t)
+
+	(defmethod :write-boot-sector (volume-size)
+		; Write the ExFat boot sector
+		; inputs
+		; volume-size = total volume size in bytes
+		(defq sector-size (get this :sector-size)
+			cluster-shift (get this :cluster-shift)
+			fat-offset (get this :fat-offset)
+			fat-length (get this :fat-length)
+			cluster-heap-offset (get this :cluster-heap-offset)
+			cluster-count (get this :cluster-count)
+			root-cluster (get this :root-dir-cluster)
+			boot-sector (str-alloc sector-size))
+
+		; Jump boot code (3 bytes)
+		(setf-code boot-sector 0 0xEB)
+		(setf-code boot-sector 1 0x76)
+		(setf-code boot-sector 2 0x90)
+
+		; File system name "EXFAT   " (8 bytes)
+		(str-cpy "EXFAT   " 0 boot-sector 3 8)
+
+		; Zeros (53 bytes) - must be zero
+		(each! 11 64 (lambda (i) (setf-code boot-sector i 0)))

```

Comment:
??? (each!) does not have range optional parameters like this ! They come at the end. See the docs for each!